### ISSUE TITLE
Haskell --cnf backend produces code with error: Illegal type synonym family application

### ISSUE THREAD

**Issue by andreasabel** [BNFC/bnfc#216]
With ghc-7.10, the code generated by the Haskell CNF backend does not compile:
```
$ bnfc --haskell --cnf -m Calc.cf 
...
$ make
```
yields
```
Compiling CnfTablesCalc    ( CnfTablesCalc.hs, CnfTablesCalc.o )

CnfTablesCalc.hs:14:10:
    Illegal type synonym family application in instance:
      [(CATEGORY, Any)]
    In the instance declaration for ‘RingP [(CATEGORY, Any)]’
```

**Comment by jyp** [comment]
Could you show some of the context in CnfTablesCalc? Maybe it's a matter of using the right GHC extension.

**Comment by andreasabel** [comment]
Here is some context (not the whole file):
```haskell
{-# LANGUAGE MagicHash, FlexibleInstances #-}
module CnfTablesCalc where
import GHC.Prim
import Control.Applicative hiding (Const)
import Algebra.RingUtils
import Parsing.Chart ()
import AbsCalc
import LexCalc
import PrintCalc
readInteger :: String -> Integer
readInteger = read
readDouble :: String -> Double
readDouble = read
instance RingP [(CATEGORY,Any)] where
  mul p a b = trav [map (app tx ty) l :/: map (app tx ty) r | (x,tx) <- a, (y,ty) <- b, let l:/:r = combine p x y]
    where trav :: [Pair [a]] -> Pair [a]
          trav [] = pure []
          trav (x:xs) = (++) <$> x <*> trav xs
          app tx ty (z,f)  = (z, f tx ty)
showAst (cat,ast) = case cat of 
      CAT_Exp -> printTree ((unsafeCoerce# ast)::Exp)
      _ -> "Unprintable category"
data CATEGORY = CAT_Exp|
                CAT_0|
                CAT_1|
                CAT_Exp1|
                CAT_2|
                CAT_3|
                CAT_Exp2|
                CAT_4|
                CAT_Integer|
                TOK_40|
                TOK_41|
                TOK_42|
                TOK_43|
                TOK_45|
                TOK_47
  deriving (Eq,Ord,Show)
```

**Comment by jyp** [comment]
Looks the `Any` type has become a type family, so the above does not work any more. I suppose that it could be wrapped into another type, like

`data OldAny = OldAny Any`

or somesuch.

**Comment by andreasabel** [comment]
What do you mean by type family here?
```
*Main> :m + Data.Monoid
*Main Data.Monoid> :info Any
newtype Any = Any {getAny :: Bool} 	-- Defined in ‘Data.Monoid’
instance Bounded Any -- Defined in ‘Data.Monoid’
instance Eq Any -- Defined in ‘Data.Monoid’
instance Ord Any -- Defined in ‘Data.Monoid’
instance Read Any -- Defined in ‘Data.Monoid’
instance Show Any -- Defined in ‘Data.Monoid’
instance Monoid Any -- Defined in ‘Data.Monoid’
```
This looks like a vanilla `newtype` to me.

**Comment by jyp** [comment]
https://hackage.haskell.org/package/base-4.10.0.0/docs/GHC-Exts.html#g:17

**Comment by andreasabel** [comment]
Ah, ok, not the `Data.Monoid.Any`.  

I have no experience with `coerce` in Haskell, haven't used that `Any` type before.

**Comment by jyp** [comment]
An alternative to using coerce/Any would be to encode the proper type using
singletons, and the dependent-haskell tricks. I'm not sure that this is a
very good idea though, as it will complicate the generated code for little
benefit.

**Comment by andreasabel** [comment]
@jyp Can you give me a date by which you will have fixed this?

**Comment by jyp** [comment]
No

**Comment by andreasabel** [comment]
Fixed by @bafain in 40cd7666a631bdd0d8db89fa1b347bff09adec2f.
